internal/cli/cmd/cluster/operations.go

```diff
@@ -131,6 +131,23 @@ func (o *OperationsOptions) addCommonFlags(cmd *cobra.Command) {
 	cmd.Flags().Lookup("dry-run").NoOptDefVal = "unchanged"
 	if o.HasComponentNamesFlag {
 		cmd.Flags().StringSliceVar(&o.ComponentNames, "components", nil, "Component names to this operations")
+		util.CheckErr(cmd.RegisterFlagCompletionFunc(
+			"components",
+			func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+				var components []string
+				if len(args) == 0 {
+					return components, cobra.ShellCompDirectiveNoFileComp
+				}
+				namespace, _, _ := f.ToRawKubeConfigLoader().Namespace()
+				dynamic, _ := f.DynamicClient()
+				cluster, _ := cluster.GetClusterByName(dynamic, args[0], namespace)
+				for _, comp := range cluster.Spec.ComponentSpecs {
+					if strings.HasPrefix(comp.Name, toComplete) {
+						components = append(components, comp.Name)
+					}
+				}
+				return components, cobra.ShellCompDirectiveNoFileComp
```

Comment:
Can extract this completion function, and call it, avoid to define it many times.